File: implementations/java-springboot/examples/src/test/java/com/example/accountservice/account/infrastructure/TransactionAutoCategorizerImplTest.java
package com.example.accountservice.account.infrastructure;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.example.accountservice.account.domain.TransactionCategory;
import com.example.accountservice.config.LlmProperties;
import java.io.IOException;
import java.net.http.HttpClient;
import java.net.http.HttpResponse;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import tools.jackson.databind.ObjectMapper;

@ExtendWith(MockitoExtension.class)
class TransactionAutoCategorizerImplTest {

    @Mock private HttpClient httpClient;

    private final ObjectMapper objectMapper = new ObjectMapper();

    private TransactionAutoCategorizerImpl categorizer;

    @BeforeEach
    void setUp() {
        LlmProperties llmProperties = new LlmProperties("http://localhost:11434", "qwen2.5:1.5b");
        categorizer = new TransactionAutoCategorizerImpl(llmProperties, objectMapper, httpClient);
    }

    @SuppressWarnings("unchecked")
    private void mockOllamaResponse(int statusCode, String content) throws Exception {
        HttpResponse<String> response = mock(HttpResponse.class);
        when(response.statusCode()).thenReturn(statusCode);
        if (content != null) {
            String body =
                    objectMapper.writeValueAsString(Map.of("message", Map.of("content", content)));
            when(response.body()).thenReturn(body);
        }
        when(httpClient.<String>send(any(), any())).thenReturn(response);
    }

    @Test
    void categorize_when_the_model_returns_a_valid_category_then_returns_it() throws Exception {
        mockOllamaResponse(200, objectMapper.writeValueAsString(Map.of("category", "FOOD")));

        TransactionCategory category = categorizer.categorize("Starbucks Gangnam", 5500);

        assertThat(category).isEqualTo(TransactionCategory.FOOD);
    }

    @Test
    void categorize_when_the_model_returns_an_out_of_taxonomy_category_then_falls_back_to_OTHER()
            throws Exception {
        mockOllamaResponse(
                200, objectMapper.writeValueAsString(Map.of("category", "NOT_A_REAL_CATEGORY")));

        TransactionCategory category = categorizer.categorize("Unknown Payee", 1000);

        assertThat(category).isEqualTo(TransactionCategory.OTHER);
    }

    @Test
    void categorize_when_the_ollama_call_fails_then_falls_back_to_OTHER_rather_than_throwing()
            throws Exception {
        when(httpClient.<String>send(any(), any()))
                .thenThrow(new IOException("connection refused"));

        TransactionCategory category = categorizer.categorize("Anything", 1000);

        assertThat(category).isEqualTo(TransactionCategory.OTHER);
    }

    @Test
    void categorize_when_ollama_responds_with_a_non_ok_status_then_falls_back_to_OTHER()
            throws Exception {
        mockOllamaResponse(500, null);

        TransactionCategory category = categorizer.categorize("Anything", 1000);

        assertThat(category).isEqualTo(TransactionCategory.OTHER);
    }
}

File: implementations/kotlin-springboot/examples/src/test/kotlin/com/example/accountservice/account/infrastructure/TransactionAutoCategorizerImplTest.kt
package com.example.accountservice.account.infrastructure

import com.example.accountservice.account.domain.TransactionCategory
import com.example.accountservice.config.LlmProperties
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import tools.jackson.module.kotlin.jacksonObjectMapper
import java.io.IOException
import java.net.http.HttpClient
import java.net.http.HttpResponse

class TransactionAutoCategorizerImplTest {
    private val httpClient = mockk<HttpClient>()
    private val objectMapper = jacksonObjectMapper()
    private lateinit var categorizer: TransactionAutoCategorizerImpl

    @BeforeEach
    fun setUp() {
        categorizer = TransactionAutoCategorizerImpl(LlmProperties("http://localhost:11434", "qwen2.5:1.5b"), objectMapper, httpClient)
    }

    private fun mockOllamaResponse(
        statusCode: Int,
        content: String?,
    ) {
        val response = mockk<HttpResponse<String>>()
        every { response.statusCode() } returns statusCode
        if (content != null) {
            val body = objectMapper.writeValueAsString(mapOf("message" to mapOf("content" to content)))
            every { response.body() } returns body
        }
        every { httpClient.send(any(), any<HttpResponse.BodyHandler<String>>()) } returns response
    }

    @Test
    fun `returns the model's category when the response is valid`() {
        mockOllamaResponse(200, objectMapper.writeValueAsString(mapOf("category" to "FOOD")))

        val category = categorizer.categorize("Starbucks Gangnam", 5500)

        assertThat(category).isEqualTo(TransactionCategory.FOOD)
    }

    @Test
    fun `falls back to OTHER rather than passing through an out-of-taxonomy category from the model`() {
        mockOllamaResponse(200, objectMapper.writeValueAsString(mapOf("category" to "NOT_A_REAL_CATEGORY")))

        val category = categorizer.categorize("Some merchant", 1000)

        assertThat(category).isEqualTo(TransactionCategory.OTHER)
    }

    @Test
    fun `falls back to OTHER rather than throwing when the ollama call fails`() {
        every { httpClient.send(any(), any<HttpResponse.BodyHandler<String>>()) } throws IOException("connection refused")

        val category = categorizer.categorize("Some merchant", 1000)

        assertThat(category).isEqualTo(TransactionCategory.OTHER)
    }

    @Test
    fun `falls back to OTHER when ollama responds with a non-ok status`() {
        mockOllamaResponse(500, null)

        val category = categorizer.categorize("Some merchant", 1000)

        assertThat(category).isEqualTo(TransactionCategory.OTHER)
    }
}
